lib/Conversion/HWToLLVM/HWToLLVM.cpp

```diff
@@ -356,12 +371,67 @@ struct HWArrayCreateOpConversion
           rewriter, op->getLoc(), typeConverter->convertType(input.getType()),
           input);
 
-      arr =
-          rewriter.create<LLVM::InsertValueOp>(op->getLoc(), arr, castInput, i);
+      arr = rewriter.create<LLVM::InsertValueOp>(op->getLoc(), arrayTy, arr,
+                                                 castInput, i);
     }
 
     rewriter.replaceOp(op, arr);
-    return success();
+  }
+};
+} // namespace
+
+namespace {
+
+static Namespace globals;
+
+/// Convert an ArrayCreateOp with constant elements to the LLVM dialect. An
+/// equivalent and initialized llvm dialect array type is generated.
+struct HWConstArrayCreateOpConversion
+    : public ConvertOpToLLVMPattern<hw::ArrayCreateOp> {
+  using ConvertOpToLLVMPattern<hw::ArrayCreateOp>::ConvertOpToLLVMPattern;
+
+  LogicalResult match(hw::ArrayCreateOp op) const override {
+    // Apply this conversion if all the array elements are constants.
+    return success(llvm::all_of(op.getOperands(), [](auto op) {
+      return isa_and_nonnull<hw::ConstantOp>(op.getDefiningOp());
+    }));
+  }
+
+  void rewrite(hw::ArrayCreateOp op, OpAdaptor adaptor,
+               ConversionPatternRewriter &rewriter) const override {
+    auto arrayTy = typeConverter->convertType(op->getResult(0).getType());
+    assert(arrayTy);
+
+    OpBuilder b(op->getParentOfType<mlir::ModuleOp>().getBodyRegion());
+
+    // Create a global region for this static array.
+    auto name = globals.newName("_array_global");
+    auto global =
+        b.create<LLVM::GlobalOp>(op->getLoc(), arrayTy, false,
+                                 LLVM::Linkage::Internal, name, Attribute(), 0);
+    Block *blk = new Block();
+    global.getInitializerRegion().push_back(blk);
+    ImplicitLocOpBuilder init(op->getLoc(), op->getContext());
+    init.setInsertionPointToStart(blk);
+
+    Value arr = init.create<LLVM::UndefOp>(op->getLoc(), arrayTy);
+    for (size_t i = 0, e = op.getInputs().size(); i < e; ++i) {
+      // Copy over the converted constant ops from the adaptor into the global
+      // region and insert them into the array.
+      Value input =
+          adaptor
+              .getInputs()[HWToLLVMEndianessConverter::convertToLLVMEndianess(
+                  op.getResult().getType(), i)];
+      auto *clone = input.getDefiningOp()->clone();
+      init.insert(clone);
+
+      Value v = clone->getResult(0);
+      arr = init.create<LLVM::InsertValueOp>(op->getLoc(), arrayTy, arr, v, i);
```

Comment:
I think it might be a good idea to add an assertion that `input.getDefiningOp()` is really a LLVM constant op because the clone does not recursively clone operands.
Also a suggestion to simplify this a bit:

```suggestion
      auto *clone = init.clone(input.getDefiningOp());
      arr = init.create<LLVM::InsertValueOp>(op->getLoc(), arr, clone->getResult(0), i);
```